test/e2e/md_rollout_test.go

```diff
@@ -0,0 +1,65 @@
+package e2e
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	bmov1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	capi_e2e "sigs.k8s.io/cluster-api/test/e2e"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("When testing MachineDeployment rolling upgrades [md-rollout]", Label("md-rollout"), func() {
+	BeforeEach(func() {
+		osType := strings.ToLower(os.Getenv("OS"))
+		Expect(osType).ToNot(Equal(""))
+		validateGlobals(specName)
+		// We need to override clusterctl apply log folder to avoid getting our credentials exposed.
+		clusterctlLogFolder = filepath.Join(os.TempDir(), "target_cluster_logs", bootstrapClusterProxy.GetName())
+
+		Logf("Removing existing BMHs from source")
+		bmhData, err := os.ReadFile(filepath.Join(workDir, bmhCrsFile))
+		Expect(err).ToNot(HaveOccurred(), "BMH CRs file not found")
+		kubeConfigPath := bootstrapClusterProxy.GetKubeconfigPath()
+		err = KubectlDelete(ctx, kubeConfigPath, bmhData, "-n", "metal3")
+		Expect(err).ToNot(HaveOccurred(), "Failed to delete existing BMHs")
```

Comment:
I would prefer checking the number of BMHs with a timeout instead of checking for the error in 1 specific moment in time. In other words let's do "Positive Testing and State Verification" whenever possible.